File: monitoring-system/server/app/services/channel_service.py
"""
Channel service — CRUD for kiosk TV channels.
"""
from ..extensions import db
from ..models.channel import Channel


def list_channels(active_only: bool = True) -> tuple[dict, int]:
    query = Channel.query
    if active_only:
        query = query.filter_by(is_active=True)
    channels = query.order_by(Channel.sort_order.asc(), Channel.id.asc()).all()
    return {"channels": [c.to_dict() for c in channels]}, 200


def create_channel(data: dict) -> tuple[dict, int]:
    channel = Channel(**data)
    db.session.add(channel)
    db.session.commit()
    return {"message": "Channel added.", "channel": channel.to_dict()}, 201


def update_channel(channel_id: int, data: dict) -> tuple[dict, int]:
    channel = Channel.query.get_or_404(channel_id)
    for key, value in data.items():
        setattr(channel, key, value)
    db.session.commit()
    return {"message": "Channel updated.", "channel": channel.to_dict()}, 200


def delete_channel(channel_id: int) -> tuple[dict, int]:
    channel = Channel.query.get_or_404(channel_id)
    db.session.delete(channel)
    db.session.commit()
    return {"message": "Channel deleted."}, 200

File: monitoring-system/server/src/services/channelService.js
import { prisma } from '../db.js';
import { HttpError } from '../utils/httpError.js';
import { channelDict } from '../utils/serialize.js';

function toRow(data) {
  return {
    name: data.name,
    stream_url: data.stream_url,
    logo_url: data.logo_url ?? '',
    stream_type: data.stream_type ?? 'hls',
    sort_order: data.sort_order ?? 0,
    is_active: data.is_active ?? true,
  };
}

export async function listChannels(activeOnly = true) {
  const where = activeOnly ? { is_active: true } : {};
  const channels = await prisma.channels.findMany({
    where,
    orderBy: [{ sort_order: 'asc' }, { id: 'asc' }],
  });
  return { channels: channels.map(channelDict) };
}

export async function createChannel(data) {
  const channel = await prisma.channels.create({ data: toRow(data) });
  return { status: 201, body: { message: 'Channel added.', channel: channelDict(channel) } };
}

export async function updateChannel(id, data) {
  const existing = await prisma.channels.findUnique({ where: { id } });
  if (!existing) throw new HttpError(404, 'Resource not found.');
  const channel = await prisma.channels.update({ where: { id }, data: toRow(data) });
  return { message: 'Channel updated.', channel: channelDict(channel) };
}

export async function deleteChannel(id) {
  const existing = await prisma.channels.findUnique({ where: { id } });
  if (!existing) throw new HttpError(404, 'Resource not found.');
  await prisma.channels.delete({ where: { id } });
  return { message: 'Channel deleted.' };
}
